service/share/share.go

```diff
@@ -0,0 +1,124 @@
+package share
+
+import (
+	"context"
+
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+
+	"github.com/celestiaorg/nmt/namespace"
+
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+// TODO(@Wondertan): We prefix real data of shares with namespaces to be able to recover them during erasure coding
+//  recovery. However, that is storage and bandwidth overhead(8 bytes per each share) which we can avoid by getting
+//  namespaces from CIDs stored in IPLD NMT Nodes, instead of encoding namespaces in erasure coding.
+//
+// TODO(@Wondertan): Ideally, Shares structure should be defined in a separate repository with
+//  rsmt2d/nmt/celestia-core/celestia-node importing it. rsmt2d and nmt are already dependent on the notion of "share",
+//  so why shouldn't we have a separated and global type for it to avoid the type mess with defining own share type in
+//  each package.
+//
+// Share is
+type Share = namespace.PrefixedData8
+
+// Service provides as simple interface to access any DataSquare/Block Share on the network.
+//
+// All Get methods follow the following flow:
+// 	* Check local storage for the requested Share.
+// 		* If exist
+// 			* Load from disk
+//			* Return
+//  	* If not
+//  		* Find provider on the network
+//      	* Fetch the Share from the provider
+//			* Store the Share
+//			* Return
+type Service interface {
+	// GetShare loads a Share committed to the given DataAvailabilityHeader by its Row anc Col coordinates in the
+	// erasures data square or block.
+	GetShare(ctx context.Context, dah header.DataAvailabilityHeader, row, col int) (Share, error)
+
+	// GetShares loads all the Shares committed to the given DataAvailabilityHeader as a 2D array/slice.
+	// It also optimistically executes erasure coding recovery.
+	GetShares(context.Context, header.DataAvailabilityHeader) ([][]Share, error)
+
+	// GetSharesByNamespace loads all the Shares committed to the given DataAvailabilityHeader as a 1D array/slice.
+	GetSharesByNamespace(context.Context, header.DataAvailabilityHeader, namespace.ID) ([]Share, error)
+
+	// Starts the Service.
+	Start(context.Context) error
+
+	// Stops the Service.
+	Stop(context.Context) error
+}
+
+// NewService creates new basic share.Service.
+func NewService(dag format.DAGService) Service {
+	return &service{
+		dag: dag,
+	}
+}
+
+type service struct {
+	dag format.DAGService
+	// session is dag sub-session that applies optimization for fetching/loading related nodes, like shares
+	// prefer session over dag for fetching nodes.
+	session format.NodeGetter
+	// cancel controls lifecycle of the session
+	cancel context.CancelFunc
+}
+
+func (s *service) Start(context.Context) error {
```

Comment:
Will context be used later once more code is implemented? Or is this interface a requirement, and requires this parameter to be there? Also, you're making a new context in the function, do you just throw it away?

Reply:
> Will context be used later once more code is implemented?

@adlerjohn, No. In the case of the share.Service, we don't need the contexts in Start and Stop, but we keep em to follow the Service pattern. 

Although, this may change as more logic is implemented in share.Service, but I don't see rn what would it be. In simple words, these contexts are needed to prevent random Node stalls due to blocking actions happening in Start/Stop. Once a long block or even deadlock happens, the context(if properly implemented) will prevent it and the returned error or log will report the issue. For share.Service we don't have any blocking operations on Start/Stopped, thus there is no context usage. 

>  Also, you're making a new context in the function, do you just throw it away?

Kinda. It is still stored in memory as the cancel func we store is kept in memory and has a ref to the ctx so we store it indirectly. In this case, we care only about canceling the context and freeing the resources allocated for the dag session.